### ISSUE TITLE
Rustdoc ignores traits implemented inside of a function

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust#52545]
After #51952 landed, I attempted to make Diesel compatible with recent nightlies by following the advice of wrapping the output in a function rather than a module. However, we are now unable to build our documentation. The compiler complains that every derived trait within Diesel is not implemented. This only occurs when running `cargo doc`, the crate otherwise compiles fine. I have tested with 1.26.1, stable, and the most recent nightly.

**Comment by oli-obk** [comment]
@GuillaumeGomez can't you do the same trick you did for `dead_code` and `unused_import`? Or is this an error?

**Comment by QuietMisdreavus** [comment]
@Mark-Simulacrum Not sure the reasoning behind tagging things as regressions - this bug has been here for all time, it's just that the hygiene change has brought this to light. Rustdoc has never looked inside functions to document things, and the introduction of `everybody-loops` made that impossible anyway. To fix this, two things need to happen:

1. `everybody-loops` needs to be revamped to somehow exclude trait impls from being removed. Alternately, we can stop doing `everybody-loops` in rustdoc, but I don't think it's just there for performance. IIRC it was introduced to rustdoc when `#[doc(cfg)]` was, to prevent trying to resolve things on the wrong platform.
2. Rustdoc needs to start scanning inside functions for impl blocks. I fear this can start recursing arbitrarily deep, though - imagine a function with an impl, that contains a function with another impl, etc. However, i imagine in practice it won't be a problem.

At some point either rustdoc or `everybody-loops` needs to filter out types that are only in-scope for that function - rustdoc may already have an easier time at this, though, since it can just treat items declared inside functions to be the equivalent of `#[doc(hidden)]`. (Meaning that you'd need to use `#[doc(no_hidden)]` to reveal trait impls inside functions? Seems legit to me.)

**Comment by Mark-Simulacrum** [comment]
I tag things as a regression mostly so that we can keep an eye on them; if we believe this isn't one we can of course untag. However, my reasoning in this case was that @sgrif indicated that Diesel changed in behavior, which feels like a regression to me, though I may have misunderstood.

**Comment by ollie27** [comment]
This looks like a duplicate of  #36922 / #41480.

**Comment by pietroalbini** [comment]
Marked as a `stable-to-stable` regression since "I have tested with 1.26.1, stable, and the most recent nightly.".

**Comment by RReverser** [comment]
This is a duplicate of my earlier report at https://github.com/rust-lang/rust/issues/41480 I believe.

**Comment by QuietMisdreavus** [comment]
@sgrif Is the version of diesel that failed to document published somewhere? Is it on crates.io or pushed to GitHub? I'm working on a fix that will at least allow it to create docs (even if the traits don't show up) but i want to make sure the original problem gets fixed.

**Comment by sgrif** [comment]
The master branch

**Comment by QuietMisdreavus** [comment]
If i can get the fixed version to compile in the first place, the fix to make the traits appear is next in line. For now, i want to make sure i can properly generate files for a crate that properly depends on the impls appearing in those positions before it can compile.

**Comment by sgrif** [comment]
Just to make clear the importance of this from my point of view -- The change which forced us to move all our trait impls into functions is now in beta. There is currently no way for us to work on 1.29 without losing the ability to build our documentation, which isn't a tradeoff we're willing to make. I'm hoping the milestone means this will be fully resolved before then. (Thank you @QuietMisdreavus for working on this)

**Comment by QuietMisdreavus** [comment]
I've opened https://github.com/rust-lang/rust/pull/53162 to make those trait impls show up in documentation.
